### ISSUE TITLE
rmsprop causing strange loss of accurracy part way through training

### ISSUE THREAD

**Issue by genixpro** [torch/optim#99]
I've been using Adagrad normally, but I decided to try Rmsprop to see if it improves accuracy. In our tests, rmsprop seemed to converge faster and to a higher maximum, so we started training our big models using it. However, I have noticed something strange that happens during training. It seems like randomly, the accuracy will suddenly precipitously drop, with loss suddenly shooting up. Sometimes I've even seen "infinity" in our testing results when this happens - as if one of the model parameters got accidentally changed to infinity, causing a cascade of failed calculations. See these results:

This is one of the first rmsprop runs:

Iteration	Time	Training Loss	Training Loss / Sequence Length	Testing Loss	Testing # Correct	Testing # Wrong	Testing # Total	Accurracy
1000	2759	871.7442105	0.8717442105	206494.6584	98616	50541	149157	66.12
2000	3193	134.9663844	0.1349663844	42419.8451	137914	11243	149157	92.46
3000	3111	103.6525763	0.1036525763	36705.87796	141171	7986	149157	94.65
4000	2971	64.08356104	0.06408356104	30939.31719	143409	5748	149157	96.15
5000	2823	51.6954515	0.0516954515	34053.09719	143495	5662	149157	96.20
6000	3156	39.32926663	0.03932926663	32059.19335	144131	5026	149157	96.63
7000	3183	126072.6807	126.0726807	inf	98386	50771	149157	65.96
8000	3018	10867.14331	10.86714331	1790859.708	104428	44729	149157	70.01
9000	3332	956.7539489	0.9567539489	164310.9163	115827	33330	149157	77.65
10000	2905	923.2978352	0.9232978352	135053.3021	116758	32399	149157	78.28
11000	3147	619.5982058	0.6195982058	134692.2995	118003	31154	149157	79.11

Notice what happens around iteration 7000. At first I thought it was a rare hardware issue, but then a different model did the same thing:

Iteration	Time	Training Loss	Testing Loss	Testing # Correct	Testing # Wrong	Testing # Total	Accurracy
1000	3032	1.998393671	3.460828	8220	140937	149157	5.51
2000	3321	1.506352061	1.13135852	106180	42977	149157	71.19
3000	3389	0.6526988754	0.6081444923	126793	22364	149157	85.01
4000	3382	0.4032474733	0.4583896942	131588	17569	149157	88.22
5000	3075	2.197617545	17.48262351	60603	88554	149157	40.63

In this second example, I can even see the point where the loss starts shooting up in the logs. It doesn't appear to be instantaneous - perhaps an error is made in one iteration that slowly cascades until it affects everything. 

decayed learning rate by a factor 0.97 to 0.01825346	
iteration 4400/17090000, seq_length = 500, loss = 0.38249470, gradnorm = 8.0499e+01. Time Elapsed: 3280 seconds	
iteration 4450/17090000, seq_length = 500, loss = 0.37212085, gradnorm = 2.9393e+02. Time Elapsed: 3426 seconds	
iteration 4500/17090000, seq_length = 500, loss = 0.36586265, gradnorm = 8.7689e+01. Time Elapsed: 3288 seconds	
iteration 4550/17090000, seq_length = 500, loss = 0.35865728, gradnorm = 5.4034e+01. Time Elapsed: 3416 seconds	
decayed learning rate by a factor 0.97 to 0.0177058562	
iteration 4600/17090000, seq_length = 500, loss = 0.40036575, gradnorm = 7.8565e+01. Time Elapsed: 3327 seconds	
iteration 4650/17090000, seq_length = 500, loss = 0.42660431, gradnorm = 2.2500e+02. Time Elapsed: 3309 seconds	
iteration 4700/17090000, seq_length = 500, loss = 0.49915671, gradnorm = 4.2741e+03. Time Elapsed: 3237 seconds	
iteration 4750/17090000, seq_length = 500, loss = 0.86534878, gradnorm = 3.5756e+03. Time Elapsed: 3251 seconds	
decayed learning rate by a factor 0.97 to 0.017174680514	
iteration 4800/17090000, seq_length = 500, loss = 1.24005108, gradnorm = 4.3706e+03. Time Elapsed: 3232 seconds	
iteration 4850/17090000, seq_length = 500, loss = 1.22130984, gradnorm = 5.6758e+03. Time Elapsed: 3117 seconds	
iteration 4900/17090000, seq_length = 500, loss = 6.12171381, gradnorm = 9.2302e+03. Time Elapsed: 3232 seconds	
iteration 4950/17090000, seq_length = 500, loss = 11.80134205, gradnorm = 9.0186e+03. Time Elapsed: 3029 seconds	
decayed learning rate by a factor 0.97 to 0.01665944009858	
iteration 5000/17090000, seq_length = 500, loss = 17.11424646, gradnorm = 6.3805e+03. Time Elapsed: 3075 seconds	

You can see loss going down, and then it starts going up again slowly, which isn't totally unusual. But then it quickly spikes and never recovers! We didn't see any "infinity's" in this run but the same curious sudden change in loss is visible.

Does anyone have any insight into what might be happening? I haven't ever seen something like this when using Adagrad - only with the models that we train using rmsprop.

**Comment by simopal6** [comment]
Same thing here, see attached picture.
![rmsprop_issue](https://cloud.githubusercontent.com/assets/4355487/14713588/f40b42c4-07e1-11e6-9677-b4369a001f83.png)

**Comment by Kaixhin** [comment]
I'm not sure if this is what is causing your issue but [line 52](https://github.com/torch/optim/blob/master/rmsprop.lua#L52) looks a little odd to me - usually epsilon should be added inside the square root. I've also seen other implementations use a default of 1e-6 instead of 1e-8.

**Comment by andreaskoepf** [comment]
Could somebody who can reliable reproduce the problem please try to use a higher epsilon value, e.g. something between 0.001 and 0.1? This should limit the maximum gain to 1000x or 10x.

Especially near a local minimum the gradient becomes small.. Since rmsprop scales the gradient the resulting gradient step might simply become too large.

@Kaixhin Could you explain why you would prefer to add epsilon before the sqrt-operation? I saw in your [rmspropm.lua](https://github.com/Kaixhin/Atari/blob/master/modules/rmspropm.lua#L8) you are using a relatively high epsilon value of `0.01`. Has this resolved the issue for you?

**Comment by andreaskoepf** [comment]
@Kaixhin it is indeed a bit inconsistent that [adagrad](https://github.com/torch/optim/blob/master/adagrad.lua#L48) adds a constant value of 1e-10 after the sqrt while [adadelta](https://github.com/torch/optim/blob/master/adadelta.lua#L47) does it before the sqrt operation.

In the nice blog post [An overview of gradient descent optimization algorithms](http://sebastianruder.com/optimizing-gradient-descent/) by @sebastianruder the epsilon is always added before taking the sqrt.

**Comment by Kaixhin** [comment]
@andreaskoepf My bad, I can't think of a reason to use one over the other - either way prevents a divide by zero. But obviously putting it inside or outside changes the value of epsilon that should be used. My reference of using an epsilon of 1e-6 and inside the square root is [Lasagne](https://github.com/Lasagne/Lasagne/blob/master/lasagne/updates.py#L400). I don't know how important this value is to optimisation practically.

As for [rmspropm](https://github.com/Kaixhin/Atari/blob/master/modules/rmspropm.lua) it is actually rmsprop with momentum, as introduced in the Graves paper referenced. I've never experimented with changing any of the default values or replacing it with the original rmsprop.

**Comment by Kaixhin** [comment]
@andreaskoepf If you have some tests where you observe Torch's rmsprop acting strangely, perhaps try using the formula/values from Lasagne. If that works then we should submit a PR to patch this.

**Comment by denis-bz** [comment]
On epsilon inside or outside the sqrt:
  x / sqrt( x^2 + r^2 )
is a smooth approximation to sign(x), ~ 0.9 at x = 2 r, a sigmoid --
~ x / r  for small x, +- 1 for large.

(Does magnifying tiny values make sense ? It's anti soft-threshold,
intuition anyone ?)
